### PR TITLE
Simplify internals

### PR DESCRIPTION
**Pull request by travisbrown**
I decided to take a stab at this on the plane yesterday, and it works out pretty nicely. The key is now that we have a viable `NonEmptyVector`, we can replace this signature:

```scala
final def onChunk(h1: E, h2: E, t: Vector[E]): F[Step[F, E, Step[F, E, A]]]
```

With this:

```scala
final def onChunk(h: E, t: NonEmptyVector[E]): F[Step[F, E, Step[F, E, A]]]
```

And we get less breaking down and building up of vectors in some situations.

I've also reworked the `Done` representation a bit so that it doesn't need `Input` at all, and have rearranged some of the `Step` subclasses.

---

Update: I want to move forward with this simplification, but as originally written it fails on 2.10 because of a bug related to value classes (specifically SI-6260). I've worked around this by adding a simple compatibility layer that provides an alternate `NonEmptyVector` implementation for 2.10 while using Cats's value class-based implementation on 2.11. (My first attempt used a `OneAnd`-based implementation for 2.10, but that led to strange JVM failures.)

What do people think about this approach? It makes 2.10 a bit of a second-class citizen, but the functionality is the same, and given that 2.12 is just around the corner and 2.10 support isn't long for this world, I'm not too worried about a temporary workaround like this.

### CODE REVIEW

**Comment by travisbrown** [comment]
Damn it, running into SI-6260 on 2.10. I hate value classes. This might have to wait until we drop 2.10 support for this change.

File: core/src/main/scala/io/iteratee/Enumeratee.scala [hunk 1]
```diff
@@ -90,23 +95,25 @@ final object Enumeratee extends EnumerateeInstances {
...
-            F.map(step.feedChunk(h1, h2, t))(loop(remaining - (t.size + 2).toLong))
+            F.map(step.feedChunk(h, t))(loop(remaining - (t.toVector.size + 1).toLong))
           } else {
-            (h1 +: h2 +: t).splitAt(remaining.toInt) match {
-              case (Vector(), nh1 +: nh2 +: nt) => F.pure(Step.doneWithLeftoverInput(step, Input.chunk(nh1, nh2, nt)))
-              case (Vector(nh), nt) => F.map(step.feedEl(nh))(Step.doneWithLeftoverInput(_, Input.fromVectorUnsafe(nt)))
-              case (nh1 +: nh2 +: nt1, nt2) => if (nt2.isEmpty) {
-                F.map(step.feedChunk(nh1, nh2, nt1))(loop(remaining - (t.size + 2).toLong))
+            (h +: t.toVector).splitAt(remaining.toInt) match {
+              case (Vector(), left) => F.pure(Step.doneWithLeftovers(step, left))
```
Comment (johnynek) [inline core/src/main/scala/io/iteratee/Enumeratee.scala]: I wonder if `if` with `size` checks would be faster here.

**Comment by johnynek** [comment]
generally seems good to me. I do wonder about perf impact of pattern matches on `Vector`.

File: core/src/main/scala/io/iteratee/internal/Step.scala [hunk 2]
```diff
@@ -360,16 +398,16 @@ final object Step { self =>
...
 
   private[this] final class TakeCont[F[_], E](acc: Vector[E], n: Int)(implicit F: Applicative[F])
-    extends PureCont.WithValue[F, E, Vector[E]](acc) {
-    final def onEl(e: E): Step[F, E, Vector[E]] = if (n == 1) done(acc :+ e) else new TakeCont(acc :+ e, n - 1)
-    final def onChunk(h1: E, h2: E, t: Vector[E]): Step[F, E, Vector[E]] = {
-      val v = h1 +: h2 +: t
+      extends PureCont.WithValue[F, E, Vector[E]](acc) {
+    final def feedElPure(e: E): Step[F, E, Vector[E]] = if (n == 1) done(acc :+ e) else new TakeCont(acc :+ e, n - 1)
+    final def feedChunkPure(h: E, t: NonEmptyVector[E]): Step[F, E, Vector[E]] = {
+      val v = h +: t.toVector
```
Comment (johnynek) [inline core/src/main/scala/io/iteratee/internal/Step.scala]: shame to build `v` just to take it apart.

File: core/src/main/scala/io/iteratee/internal/Step.scala [hunk 3]
```diff
@@ -384,14 +422,14 @@ final object Step { self =>
...
-    extends PureCont.WithValue[F, E, Vector[E]](acc) {
-    final def onEl(e: E): Step[F, E, Vector[E]] =
-      if (p(e)) new TakeWhileCont(acc :+ e, p) else new WithLeftovers(acc, Input.el(e))
-    final def onChunk(h1: E, h2: E, t: Vector[E]): Step[F, E, Vector[E]] = {
-      val (before, after) = (h1 +: h2 +: t).span(p)
+      extends PureCont.WithValue[F, E, Vector[E]](acc) {
+    final def feedElPure(e: E): Step[F, E, Vector[E]] =
+      if (p(e)) new TakeWhileCont(acc :+ e, p) else Done(acc, Vector(e))
+    final def feedChunkPure(h: E, t: NonEmptyVector[E]): Step[F, E, Vector[E]] = {
+      val (before, after) = (h +: t.toVector).span(p)
```
Comment (johnynek) [inline core/src/main/scala/io/iteratee/internal/Step.scala]: can we not build the intermediate vector and instead just call `.span` on t?

File: core/src/main/scala/io/iteratee/internal/Step.scala [hunk 4]
```diff
@@ -402,17 +440,15 @@ final object Step { self =>
...
-        val len = t.size + 2
+      final def feedElPure(e: E): Step[F, E, Unit] = drop(n - 1)
+      final def feedChunkPure(h: E, t: NonEmptyVector[E]): Step[F, E, Unit] = {
+        val len = t.toVector.size + 1
 
         if (len <= n) drop(n - len) else {
-          val dropped = (h1 +: h2 +: t).drop(n)
-
-          new WithLeftovers((), Input.fromVectorUnsafe(dropped))
+          Done((), (h +: t.toVector).drop(n))
```
Comment (johnynek) [inline core/src/main/scala/io/iteratee/internal/Step.scala]: can't we call `drop(n-1)` on t if `n > 0`?

File: core/src/main/scala/io/iteratee/internal/Step.scala [hunk 5]
```diff
@@ -425,20 +461,19 @@ final object Step { self =>
...
 
   private[this] final class DropWhileCont[F[_], E](p: E => Boolean)(implicit F: Applicative[F])
-    extends PureCont[F, E, Unit] {
-    final def onEl(e: E): Step[F, E, Unit] = if (p(e)) dropWhile(p) else new WithLeftovers((), Input.el(e))
-    final def onChunk(h1: E, h2: E, t: Vector[E]): Step[F, E, Unit] = {
-      val after = (h1 +: h2 +: t).dropWhile(p)
+      extends PureCont[F, E, Unit] {
+    final def feedElPure(e: E): Step[F, E, Unit] = if (p(e)) dropWhile(p) else Done((), Vector(e))
+    final def feedChunkPure(h: E, t: NonEmptyVector[E]): Step[F, E, Unit] = {
+      val after = (h +: t.toVector).dropWhile(p)
```
Comment (johnynek) [inline core/src/main/scala/io/iteratee/internal/Step.scala]: similar construction to deconstruct pattern.

**Comment by travisbrown** [comment]
Superseded by #141 (and #144).

**Pull request closed without merging**